services/QuillLMS/deploy.sh

```diff
@@ -71,6 +68,41 @@ case $1 in
     exit 1
 esac
 
+promote() {
+  git fetch origin
+  local latest_prod_release_sha=`heroku releases -a empirical-grammar | grep Deploy | cut -d ' ' -f 4 | head -n 1`
+  local promotion_candidate_sha=`heroku releases -a $HEROKU_APP | grep Deploy | cut -d ' ' -f 4 | head -n 1`
+
+  # if promotion candidate is not a commit descendent of the current production slug, abort
+  git merge-base --is-ancestor $latest_prod_release_sha $promotion_candidate_sha
```

Comment:
I iterated on the safety mechanism a couple times, first with timestamps, then with git tags. I believe that directly checking for ancestry is the cleanest and most precise method. 

We are able to use this method now because our prod branch, now called `develop`, is fast-forward (no merges, only squashes). But even in the case where a conventional merge to `develop` happens (someone misclicks the `Merge` button and does not specify `squash`, this method fails gracefully: the promotion will not happen, and the deployer will be nudged to promote a different slug, or deploy conventionally.

Reply:
Do we want to disable merge commits in the UI flow?

Reply:
We could; there's a conceptual elegance to that that is appealing. I'm ultra-cautious of adding checks that could prevent us from getting new code to production in an emergency scenario or in some weird edge case I haven't anticipated, so I opted not to do that here. To me, the minor annoyance of getting my promotion rejected by the CLI is acceptable.